### DISCUSSION TITLE
Should ZFS members be detected by lsblk?

### DISCUSSION THREAD

**Discussion by TheLillo** [openzfs/zfs#14348] [Ideas]
Is there a way to use lsblk in a Linux system on ZFS that shows information like mountpoint, members etc?

**Comment by darkralts** [comment]
i want to know

**Comment by rincebrain** [comment]
That's kind of like asking for the mountpoint of an LVM volume group - the answer may be, depending on what you're doing with it, "none", "one" ,or "many".

That said, I've never looked at what `lsblk` reports for LVM and similar, so if it can do it for that, conceivably one could convince it to do that for ZFS.

**Reply by LunarLambda** [reply]
lsblk can report LVs and their mountpoints, nested underneath their backing devices (PVs).

Here's an example I pulled from a server:

```
NAME                 MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINTS
sda                    8:0    0  320G  0 disk
├─sda1                 8:1    0    2M  0 part
├─sda2                 8:2    0    1G  0 part  /boot
└─sda3                 8:3    0  319G  0 part
  └─sda3_crypt       254:0    0  319G  0 crypt
    ├─vg1-swap     254:1    0  7.4G  0 lvm   [SWAP]
    ├─vg1-root     254:2    0  7.4G  0 lvm   /
    ├─vg1-machines 254:3    0  220G  0 lvm   /var/lib/machines
```

If a VG has multiple PVs, I believe `lsblk` will show the LVs underneath each PV in the default view, which can lead to some duplicated output, in a similar way as it does for `mdraid` devices.

**Reply by amotin** [reply]
In the examples of mdraid and lvm it seems to be a block layer on top of block layer, and then for the last block layer reported its mount point for traditional file systems.  I don't know how it gets the mount points, but I guess not by reading the disks, but somehow reading superblock information form kernel to associate it with the disks. Whether it can be done from ZFS, I don't know, but considering that many pools have many datasets and many vdevs, I am not sure the output would end up useful.

**Reply by LunarLambda** [reply]
I think at minimum for the kernel to associate some sort of blockdev with a mountpoint (and thus have it appear in lsblk) it would need to not show up as `nodev` in `/proc/filesystems`

EDIT: seems to be `mtab` / `/proc/mounts` based somehow?

https://github.com/util-linux/util-linux/blob/master/misc-utils/lsblk-mnt.c
